File: frontend/src/utils/growthStage.ts
export type GrowthStage = "seed" | "growing" | "harvest" | "unknown";

export interface GrowthStageInfo {
  stage: GrowthStage;
  icon: string;
  message: string;
}

/**
 * Determine the growth stage for a holding based on metrics.
 *
 * When `daysHeld` is null/undefined (no real acquisition date on record —
 * see #7220), this deliberately returns the distinct "unknown" stage rather
 * than falling through to "seed": a missing date is not evidence of a new
 * position, and every holding silently defaulting to the same stage is the
 * exact bug this guards against.
 */
export function getGrowthStage({
  daysHeld,
  currentPrice,
  targetPrice,
}: {
  daysHeld?: number | null;
  currentPrice?: number | null;
  targetPrice?: number | null;
}): GrowthStageInfo {
  if (
    targetPrice != null &&
    currentPrice != null &&
    currentPrice >= targetPrice
  ) {
    return {
      stage: "harvest",
      icon: "🍾",
      message: "Target met – consider selling.",
    };
  }
  if (daysHeld == null) {
    return {
      stage: "unknown",
      icon: "❔",
      message: "No acquisition date on record – growth stage unknown.",
    };
  }
  if (daysHeld >= 180) {
    return {
      stage: "harvest",
      icon: "🍾",
      message: "Long-term hold – review position.",
    };
  }
  if (daysHeld >= 30) {
    return {
      stage: "growing",
      icon: "🌿",
      message: "Growing – monitor performance.",
    };
  }
  return {
    stage: "seed",
    icon: "🌱",
    message: "New position – give it time to grow.",
  };
}

export default getGrowthStage;

File: backend/utils/growth_stage.py
"""Determine growth stage for holdings."""
from __future__ import annotations
from typing import Optional, Dict

StageInfo = Dict[str, str]

def get_growth_stage(
    days_held: Optional[int] = None,
    current_price: Optional[float] = None,
    target_price: Optional[float] = None,
) -> StageInfo:
    """Return a dictionary describing the growth stage of a holding."""
    if (
        target_price is not None
        and current_price is not None
        and current_price >= target_price
    ):
        return {
            "stage": "harvest",
            "icon": "🍾",
            "message": "Target met – consider selling.",
        }
    if days_held is not None and days_held >= 180:
        return {
            "stage": "harvest",
            "icon": "🍾",
            "message": "Long-term hold – review position.",
        }
    if days_held is not None and days_held >= 30:
        return {
            "stage": "growing",
            "icon": "🌿",
            "message": "Growing – monitor performance.",
        }
    return {
        "stage": "seed",
        "icon": "🌱",
        "message": "New position – give it time to grow.",
    }
